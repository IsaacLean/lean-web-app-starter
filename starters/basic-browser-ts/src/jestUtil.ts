/**
 * Clear the root element.
 * This is intended to completely empty the root element's children between tests
 * so test results can be consistent.
 * @param rootEl Optionally pass in the rootEl to be cleared, otherwise a query for the element with the root ID will be made
 */
export const clearRootEl = (rootEl?: HTMLElement) => {
    const r = rootEl ? rootEl : document.getElementById('root')

    if (!r) {
        throw new Error('HTML element with an ID of "root" was not found.')
    }

    r.innerHTML = ''
}

/**
 * Setup the root element.
 * The root div element is intended for other elements to mount to for testing.
 * @returns The new root element
 */
export const setupRootEl = () => {
    const rootEl = document.createElement('div')
    rootEl.setAttribute('id', 'root')
    document.body.replaceChildren(rootEl)

    return rootEl
}
