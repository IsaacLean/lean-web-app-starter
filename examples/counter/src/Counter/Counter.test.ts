import { screen } from "@testing-library/dom";
import userEvent from "@testing-library/user-event";

import Counter from ".";
import { clearRootEl, setupRootEl } from "../jestUtil";

describe("Counter element", () => {
  beforeAll(() => setupRootEl());

  afterEach(() => clearRootEl());

  /**
   * Create the counter element and mount it to the root element.
   * @returns
   */
  function setupTest() {
    const rootEl = document.getElementById("root");

    if (!rootEl) {
      throw new Error('HTML element with an ID of "root" was not found.');
    }

    const c = new Counter();
    const counterElement = c.createElement();
    rootEl.appendChild(counterElement);

    return counterElement;
  }

  it("matches snapshot", () => {
    expect(setupTest()).toMatchSnapshot();
  });

  it("increments count when + button is clicked", async () => {
    expect.assertions(4);

    const user = userEvent.setup();

    setupTest();

    // Expect counter to start at 0
    expect(screen.queryByText(0)).toBeInTheDocument();
    expect(screen.queryByText(1)).not.toBeInTheDocument();

    // Click the + button once
    await user.click(screen.getByRole("button", { name: /increment count/i }));

    // Expect counter to have been incremented to 1
    expect(screen.queryByText(1)).toBeInTheDocument();
    expect(screen.queryByText(0)).not.toBeInTheDocument();
  });

  it("decrements count when - button is clicked", async () => {
    expect.assertions(4);

    const user = userEvent.setup();

    setupTest();

    // Expect counter to start at 0
    expect(screen.queryByText(0)).toBeInTheDocument();
    expect(screen.queryByText(-1)).not.toBeInTheDocument();

    // Click the - button once
    await user.click(screen.getByRole("button", { name: /decrement count/i }));

    // Expect counter to have been decremented to -1
    expect(screen.queryByText(-1)).toBeInTheDocument();
    expect(screen.queryByText(0)).not.toBeInTheDocument();
  });

  it("changes count to force count value when force button is clicked", async () => {
    expect.assertions(4);

    const user = userEvent.setup();

    setupTest();

    // Expect counter to start at 0
    expect(screen.queryByText(0)).toBeInTheDocument();
    expect(screen.queryByText(5)).not.toBeInTheDocument();

    await user.type(screen.getByRole("spinbutton"), "5");
    await user.click(screen.getByRole("button", { name: /force!/i }));

    // Expect counter to have been forced to 5
    expect(screen.queryByText(5)).toBeInTheDocument();
    expect(screen.queryByText(0)).not.toBeInTheDocument();
  });

  it("disables force button when force count input is empty", async () => {
    expect.assertions(4);

    const user = userEvent.setup();

    setupTest();

    // Expect force input to start empty
    const inputForceCount = screen.getByRole("spinbutton");
    expect(inputForceCount).toHaveValue(null);

    // Expect force button to start disabled
    const btnForceCount = screen.getByRole("button", { name: /force!/i });
    expect(btnForceCount).toBeDisabled();

    await user.type(inputForceCount, "5");

    // Expect force button to become enabled
    expect(btnForceCount).toBeEnabled();

    await user.clear(inputForceCount);

    // Expect force button to become disabled again
    expect(btnForceCount).toBeDisabled();
  });

  it("resets count when reset button is clicked", async () => {
    expect.assertions(6);

    const user = userEvent.setup();

    setupTest();

    // Expect counter to start at 0
    expect(screen.queryByText(0)).toBeInTheDocument();
    expect(screen.queryByText(3)).not.toBeInTheDocument();

    // Click the + button 3x
    await user.tripleClick(
      screen.getByRole("button", { name: /increment count/i }),
    );

    // Expect counter to have been incremented to 3
    expect(screen.queryByText(3)).toBeInTheDocument();
    expect(screen.queryByText(0)).not.toBeInTheDocument();

    // Click the reset button
    await user.click(screen.getByRole("button", { name: /reset/i }));

    // Expect counter to return to 0
    expect(screen.queryByText(0)).toBeInTheDocument();
    expect(screen.queryByText(3)).not.toBeInTheDocument();
  });
});

describe("Counter class", () => {
  it("defaults counter to 0", () => {
    const c = new Counter();
    expect(c.getCount()).toBe(0);
  });

  it("defaults counter to 2 when 2 is passed into the constructor", () => {
    const c = new Counter(2);
    expect(c.getCount()).toBe(2);
  });

  it("sets the count to 1 when setCount(1) is called", () => {
    const c = new Counter();
    c.setCount(1);
    expect(c.getCount()).toBe(1);
  });
});
