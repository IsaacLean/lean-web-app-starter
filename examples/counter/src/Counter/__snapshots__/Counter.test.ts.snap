// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter element matches snapshot 1`] = `
<div
  class="align-items-center container d-flex justify-content-center h-100"
>
  <div
    class="card"
  >
    <div
      class="card-header"
    >
      Counting App
    </div>
    <div
      class="card-body d-flex flex-column align-items-center px-5"
    >
      <p
        class="lead"
      >
        Count is currently...
      </p>
      <p
        class="display-1"
      >
        0
      </p>
      <div
        class="column-gap-3 d-flex w-100"
      >
        <button
          class="btn btn-success w-50"
        >
          <i
            class="bi bi-dash-circle"
            style="font-size: 1.5rem;"
          >
            <span
              class="sr-only"
            >
              Decrement Count
            </span>
          </i>
        </button>
        <button
          class="btn btn-success w-50"
        >
          <i
            class="bi bi-plus-circle-fill"
            style="font-size: 1.5rem;"
          >
            <span
              class="sr-only"
            >
              Increment Count
            </span>
          </i>
        </button>
      </div>
      <hr
        class="w-100"
      />
      <div
        class="input-group"
      >
        <input
          class="form-control"
          placeholder="Force count to be..."
          type="number"
        />
        <button
          class="btn btn-danger"
          disabled=""
        >
          Force!
        </button>
      </div>
      <hr
        class="w-100"
      />
      <button
        class="btn btn-primary w-100"
      >
        Reset
      </button>
    </div>
  </div>
</div>
`;
